### PR TITLE
Add support for the `log` crate

### LINKED ISSUE #865: use log::debug macro for debug function in .60 language
It would be nice if using the built-in debug function in .60 files autogenerated Rust code which used the debug macro from the [log](https://lib.rs/crates/log) crate. This would allow applications to choose the logging implementation instead of always printing directly to standard out. It would also be nice to add functions for the other log levels (trace, info, warn, error).

### PR DESCRIPTION
**Pull request by lukas-jung**
Fixes  #865

### CODE REVIEW

File: api/rs/slint/Cargo.toml [hunk 1]
```diff
@@ -49,6 +49,10 @@ backend-gl-x11 = ["i-slint-backend-selector/backend-gl-x11", "std"]
 ## Wayland window system on Unix.
 backend-gl-wayland = ["i-slint-backend-selector/backend-gl-wayland", "std"]
 
+## If enabled, calls of `debug()` in `.slint` files use to the `debug!()` macro
+## of the `log` crate instead of just `println!()`.
+log = ["dep:log"]
```
Comment (ogoffart) [inline api/rs/slint/Cargo.toml:54]: you actually don't need that, it is automatic.  And you can add the `##` before the `log`  dependency

File: api/rs/slint/lib.rs [hunk 2]
```diff
@@ -445,6 +445,18 @@ pub mod internal {
...
+
+    #[doc(hidden)]
+    pub fn debug(s: SharedString) {
+        #[cfg(feature = "log")]
+        {
+            log::debug!("{:?}", s);
+        }
+        #[cfg(not(feature = "log"))]
+        {
+            println!("{:?}", s);
```
[the remark this answers is not in the document]
Comment (lukas-jung) [reply api/rs/slint/lib.rs:457]: Then it just silently will not print if the log feature isn't activated.
I think we should throw a compile_error instead and tell the users to register a logger.

[the remark this answers is not in the document]
Comment (tronical) [reply api/rs/slint/lib.rs:457]: Would it perhaps be possible to disable the `std` feature of the `log` crate?

File: api/rs/slint/Cargo.toml (hunk 1 above)
Comment (lukas-jung) [reply api/rs/slint/Cargo.toml:54]: So should i move the documentation to the dependency section or should we let the feature entry commented out until we can use `dep:` and then uncomment it to have it listed as a feature explicitly?

Comment (ogoffart) [reply api/rs/slint/Cargo.toml:54]: move the documentation to the dependency section

File: api/rs/slint/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ogoffart) [reply api/rs/slint/lib.rs:457]: > Would it perhaps be possible to disable the std feature of the log crate?

Looks like the std feature is disabled by default anyway, so i think we're good

> Then it just silently will not print if the log feature isn't activated.

I'd say it's fine if it doesn't print.
When std is disabled, a lot of other things are not working. 
Maybe we could just panic.

Another idea would be to use i_slint_core::debug_log

**Review by ogoffart — APPROVED** [review]

File: api/rs/slint/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ogoffart) [reply api/rs/slint/lib.rs:458]: We also could keep it a compilation error if the whole `debug` was disabled in that case.

**Pull request merged**